framework/CodeInterfaces/PHISICS/MaterialParser.py

```diff
@@ -0,0 +1,157 @@
+    """
+      Transforms the dictionary of dictionaries from the XML tree to a dictionary of dictionaries formatted identically as the perturbed dictionary. 
+      @ In, None 
+      @ Out, XMLdict, dictionary, under the format {'DENSITY':{'FUEL1':{'U238':1.000}}}
+    """
+    XMLdict = {}
+    matList = []
+    isotopeList = []
+    XMLdict['density'] = {}
+    for matXML in self.root.getiterator('mat'):
+      for isotopeXML in self.root.getiterator('isotope'):
+        matList.append(matXML.attrib.get('id'))
+        isotopeList.append(isotopeXML.attrib.get('id'))
+    matList = self.unifyElements(matList)
+    isotopeList = self.unifyElements(isotopeList)
+    for i in xrange(0,len(matList)):
+      XMLdict['density'][matList[i]] = {}
+      for j in xrange(0,len(isotopeList)):
+        XMLdict['density'][matList[i]][isotopeList[j]] = {}  
+    for matXML in self.root.getiterator('mat'): 
+      for isotopeXML in matXML.findall('isotope'):
+        XMLdict['density'][matXML.attrib.get('id')][isotopeXML.attrib.get('id')] = isotopeXML.attrib.get('density')
+    return XMLdict
+  
+  def unifyElements(self,listWithRepetitions):
+    """
+      Takes a list as an entry and reduces the list into a list of unique elements. 
+      @ In, listWithRepetitions, list 
+      @ Out, listWithUniqueElements, list
+    """
+    valueSeen = set() 
+    listWithUniqueElements = [x for x in listWithRepetitions if x not in valueSeen and not valueSeen.add(x)]
+    return listWithUniqueElements
+    
+  def __init__(self, inputFiles, **pertDict):
+    """
+      @ In, inputFiles, string, Qvalues library file 
+      @ In, pertDict, dictionary, dictionary of perturbed variables
+      @ Out, None
+    """
+    self.pertDict = pertDict
+    for key, value in self.pertDict.iteritems(): 
+      self.pertDict[key] = '%.3E' % Decimal(str(value)) #convert the values into scientific values 
+    self.inputFiles = inputFiles
+    self.tree = ET.parse(self.inputFiles)
+    self.root = self.tree.getroot()
+    self.listedDict = self.fileReconstruction(self.pertDict)
+    self.printInput()
+
+  def fileReconstruction(self, deconstructedDict):
+    """
+      Converts the formatted dictionary -> {'DENSITY|FUEL1|U235':1.30, DENSITY|FUEL2|U238':4.69} 
+      into a dictionary of dictionaries that has the format -> {'DENSITY':{'FUEL1':{'U235':1.30}, 'FUEL2':{'U238':4.69}}}
+      @ In, deconstructedDict, dictionary 
+      @ Out, reconstructedDict, nested dictionary 
+    """
+    reconstructedDict           = {}
+    perturbedIsotopes           = []
+    perturbedMaterials          = []
+    perturbedPhysicalParameters = []
+    for i in deconstructedDict.iterkeys() :
+      splittedKeywords = i.split('|')
+      perturbedIsotopes.append(splittedKeywords[2])
+      perturbedMaterials.append(splittedKeywords[1])
+      perturbedPhysicalParameters.append(splittedKeywords[0])
+    for i in xrange (0,len(perturbedPhysicalParameters)):
```

Comment:
you do not need to use these fortran style loops (that are slower)

Reply:
Do you mean "change from 
`for i in xrange (0,len(perturbedPhysicalParameters)):` 
to 
`for i in range (len(perturbedPhysicalParameters)):` ", 
or do you mean that I don't need to declare the dictionaries `reconstructedDict[perturbedPhysicalParameters[i]] = {}`